extensions/wikia/CrosslinkTag/CrosslinkTagController.class.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+class CrosslinkTagController extends WikiaController {
+
+	const PARSER_TAG_NAME = 'storieslink';
+	const MAX_URLS = 4;
+	const MAX_LENGTH = 90;
+
+	protected $counter = 0;
+	protected $markers = [];
+	protected static $instance = null;
+
+	/**
+	 * @return CrosslinkTagController
+	 */
+	public static function getInstance() {
+		if ( is_null( static::$instance ) ) {
+			static::$instance = new CrosslinkTagController();
+		}
+
+		return static::$instance;
+	}
+
+	/**
+	 * Hook: used to register parser tag
+	 * @param Parser $parser
+	 * @return bool
+	 */
+	public static function onParserFirstCallInit( Parser $parser ) {
+		$parser->setHook( self::PARSER_TAG_NAME, [ static::getInstance(), 'renderTag' ] );
+
+		return true;
+	}
+
+	/**
+	 * Render tag
+	 *
+	 * @param string $content - tag content
+	 * @param array $attributes - tag attributes
+	 * @param Parser $parser - parser instance
+	 * @param PPFrame $frame - parent frame with the context
+	 * @return string
+	 */
+	public function renderTag( $content, $attributes, Parser $parser, PPFrame $frame ) {
+		$markerId = $this->getMarkerId( $parser );
+
+		$urls = explode( PHP_EOL, trim( $content ) );
+		$html = $this->app->renderView( 'CrosslinkTag', 'render', [ 'markerId' => $this->counter, 'urls' => $urls ] );
+		if ( !empty( $html ) ) {
+			$html .= $this->getJSSnippet();
+		}
+
+		$this->markers[$markerId] = $html;
+
+		return $markerId;
+	}
+
+	/**
+	 * Get marker id
+	 *
+	 * @param Parser $parser
+	 * @return string
+	 */
+	protected function getMarkerId( Parser $parser ) {
+		$this->counter++;
+		$markerId = $parser->uniqPrefix() . '-' . self::PARSER_TAG_NAME . '-' . $this->counter . Parser::MARKER_SUFFIX;
+
+		return $markerId;
+	}
+
+	/**
+	 * Hook: replace tag marker
+	 * @param Parser $parser
+	 * @param string $text
+	 * @return bool
+	 */
+	public static function onParserAfterTidy( Parser &$parser, &$text ) {
+		if ( empty( F::app()->wg->ArticleAsJson ) ) {
+			$text = static::getInstance()->replaceMarkers( $text );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Hook: replace tag marker
+	 * @param string $text
+	 * @return bool
+	 */
+	public static function onArticleAsJsonBeforeEncode( &$text ) {
+		$text = static::getInstance()->replaceMarkers( $text );
+
+		return true;
+	}
+
+	/**
+	 * Replace markers
+	 * @param string $text
+	 * @return string
+	 */
+	protected function replaceMarkers( $text ) {
+		return strtr( $text, $this->markers );
+	}
+
+	/**
+	 * Add assets to be loaded
+	 * @return string
+	 */
+	protected function getJSSnippet() {
+		$html = JSSnippets::addToStack(
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Immediately return this expression instead of assigning it to the temporary variable "$html". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1488)